tests/unit/s2n_self_talk_session_resumption_test.c

```diff
@@ -0,0 +1,265 @@
+    
+    /* To ensure we are not re-using session tickets */
+    POSIX_CHECKED_MEMSET((uint8_t *)cb_session_data, 0, MAX_TEST_SESSION_SIZE);
+
+    EXPECT_SUCCESS(s2n_session_ticket_get_data_len(ticket, &cb_session_data_len));
+    EXPECT_SUCCESS(s2n_session_ticket_get_data(ticket, cb_session_data_len, cb_session_data));
+    EXPECT_SUCCESS(s2n_session_ticket_get_lifetime(ticket, &cb_session_lifetime));
+
+    return S2N_SUCCESS;
+}
+
+static int s2n_setup_test_ticket_key(struct s2n_config *config)
+{
+    POSIX_ENSURE_REF(config);
+
+    /**
+     *= https://tools.ietf.org/rfc/rfc5869#appendix-A.1
+     *# PRK  = 0x077709362c2e32df0ddc3f0dc47bba63
+     *#        90b6c73bb50f9c3122ec844ad7c2b3e5 (32 octets)
+     **/
+    S2N_BLOB_FROM_HEX(ticket_key,
+    "077709362c2e32df0ddc3f0dc47bba63"
+    "90b6c73bb50f9c3122ec844ad7c2b3e5");
+
+    /* Set up encryption key */
+    uint64_t current_time;
+    uint8_t ticket_key_name[16] = "2016.07.26.15\0";
+    EXPECT_SUCCESS(s2n_config_set_session_tickets_onoff(config, 1));
+    EXPECT_SUCCESS(config->wall_clock(config->sys_clock_ctx, &current_time));
+    EXPECT_SUCCESS(s2n_config_add_ticket_crypto_key(config, ticket_key_name, strlen((char *)ticket_key_name),
+                    ticket_key.data, ticket_key.size, current_time/ONE_SEC_IN_NANOS));
+
+    return S2N_SUCCESS;
+}
+
+int main(int argc, char **argv)
+{
+    BEGIN_TEST();
+
+    /* Setup server config */
+    struct s2n_config *server_config = s2n_config_new();
+    EXPECT_NOT_NULL(server_config);
+    EXPECT_SUCCESS(s2n_config_set_cipher_preferences(server_config, "default_tls13"));
+    EXPECT_SUCCESS(s2n_config_set_unsafe_for_testing(server_config));
+    struct s2n_cert_chain_and_key *chain_and_key = NULL;
+    EXPECT_SUCCESS(s2n_test_cert_chain_and_key_new(&chain_and_key, S2N_ECDSA_P384_PKCS1_CERT_CHAIN,
+                                                   S2N_ECDSA_P384_PKCS1_KEY));
+    EXPECT_SUCCESS(s2n_config_add_cert_chain_and_key_to_store(server_config, chain_and_key));
+    struct s2n_cert_chain_and_key *tls12_chain_and_key = NULL;;
+    EXPECT_SUCCESS(s2n_test_cert_chain_and_key_new(&tls12_chain_and_key, S2N_DEFAULT_TEST_CERT_CHAIN, 
+                                                   S2N_DEFAULT_TEST_PRIVATE_KEY));
+    EXPECT_SUCCESS(s2n_config_add_cert_chain_and_key_to_store(server_config, tls12_chain_and_key));
+    EXPECT_SUCCESS(s2n_config_set_session_tickets_onoff(server_config, true));
+    EXPECT_SUCCESS(s2n_setup_test_ticket_key(server_config));
+    EXPECT_SUCCESS(s2n_config_set_initial_ticket_count(server_config, 1));
+
+    /* Setup TLS1.3 client config */
+    struct s2n_config *tls13_client_config = s2n_config_new();
+    EXPECT_NOT_NULL(tls13_client_config);
+    EXPECT_SUCCESS(s2n_config_set_cipher_preferences(tls13_client_config, "default_tls13"));
+    EXPECT_SUCCESS(s2n_config_set_unsafe_for_testing(tls13_client_config));
+    EXPECT_SUCCESS(s2n_config_set_session_tickets_onoff(tls13_client_config, true));
```

Comment:
Do you need to set session ticket on for the client?

Reply:
Yes, for both TLS1.3 and TLS1.2 clients. Otherwise the client won't serialize the ticket: https://github.com/aws/s2n-tls/blob/main/tls/s2n_resume.c#L167

Reply:
Should we ignore NewSessionTickets if session tickets are turned off then?

Reply:
Hmm yes, I don't see a problem with that. I'm pretty sure there's no change in behavior if we did that.

Reply:
Can you either fix that as part of this PR, or create an issue?

Reply:
As an extra precaution, can you error in that serialization method if we try to serialize with session id for TLS1.3? That's not supported yet, and should never happen.

You could also change the method that sets the callback to also call `s2n_config_set_session_tickets_onoff`

Reply:
And do we have the same problem when deserializing? What happens if a server without that flag set receives a resumption PSK?

Reply:
I added a check for this when reading resumption identities. It would error eventually, but we don't need to waste time trying to decrypt if use_tickets == 0